### PR TITLE
feat(cubit): add onTransition

### PR DESCRIPTION
**Pull request by felangel**
- feat: add `onTransition` to `cubit`

### CODE REVIEW

File: packages/cubit/lib/src/cubit.dart [hunk 1]
```diff
@@ -25,11 +27,34 @@ abstract class Cubit<T> extends Stream<T> {
...
+  void onTransition(Transition<Null, T> transition) {}
+
   /// Updates the [state] of the `cubit` to the provided [state].
   /// [emit] does nothing if the `cubit` has been closed.
   @protected
   void emit(T state) async {
     if (state == _state || _controller.isClosed) return;
+    onTransition(Transition(
+      currentState: _state,
+      event: null,
```
Comment (samandmoore) [inline packages/cubit/lib/src/cubit.dart]: What's the thinking behind the event parameter? Cubit doesn't have a concept up events, right? Is this somehow meant to support the needs of Bloc?

**Review by samandmoore — COMMENTED** [review]
Exciting progress here!

File: packages/cubit/lib/src/cubit.dart (hunk 1 above)
Comment (felangel) [reply packages/cubit/lib/src/cubit.dart]: We might be able to determine the name of the method from which the state was emitted (more on this soon) so I was thinking it made sense to leave the door open for that possibility 😛

Comment (samandmoore) [reply packages/cubit/lib/src/cubit.dart]: Gotchaaaaaa. Cool. Thanks for sharing.

File: packages/cubit/lib/src/cubit.dart [hunk 2]
```diff
@@ -47,14 +47,18 @@ abstract class Cubit<T> extends Stream<T> {
 
   /// Updates the [state] of the `cubit` to the provided [state].
   /// [emit] does nothing if the `cubit` has been closed.
+  ///
+  /// [track] can optionally be provided to determine whether the
+  /// state change should be tracked (reported to `onTransition`).
+  /// Default is `true`.
   @protected
-  void emit(T state) async {
+  void emit(T state, {bool track}) async {
```
Comment (zepfietje) [inline packages/cubit/lib/src/cubit.dart]: I'd rather default using:
```suggestion
  void emit(T state, {bool track = true}) async {
```
Currently, when calling `emit(state + 1, track: null)` this leads to the state change being tracked, which is undesirable.

File: packages/cubit/lib/src/cubit.dart [hunk 3]
```diff
@@ -47,14 +47,18 @@ abstract class Cubit<T> extends Stream<T> {
...
     if (state == _state || _controller.isClosed) return;
-    onTransition(Transition(
-      currentState: _state,
-      event: null,
-      nextState: state,
-    ));
+    if (track != false) {
```
Comment (zepfietje) [inline packages/cubit/lib/src/cubit.dart]:
```suggestion
    if (track == true) {
```

File: packages/cubit/test/cubit_test.dart [hunk 4]
```diff
@@ -48,6 +48,31 @@ void main() {
...
+        final transitions = <Transition<Null, int>>[];
+        final cubit = CounterCubit(onTransitionCallback: transitions.add);
+        await Future<void>.delayed(
+          Duration.zero,
+          () => cubit.increment(track: false),
+        );
+        await cubit.close();
+        expect(transitions, isEmpty);
+      });
+
```
Comment (zepfietje) [inline packages/cubit/test/cubit_test.dart]: Maybe add another test `is not called when track is null`.

**Review by zepfietje — CHANGES_REQUESTED** [review]
@felangel could you take a look at proposed changes.

File: packages/cubit/lib/src/cubit.dart (hunk 3 above)
Comment (tenhobi) [reply packages/cubit/lib/src/cubit.dart]: NNBD coming soon. :-) @felangel we might port this library to NNBD now. It should be backward compatible.

File: packages/cubit/lib/src/cubit.dart (hunk 2 above)
Comment (felangel) [reply packages/cubit/lib/src/cubit.dart]: Yeah I agree. The only reason I chose to go with this approach is because bloc will be extending Cubit and it felt a bit weird to have the following

@override
void emit(State state, {bool track = true}) {
  super.emit(state, track: false);
}

Also it’s not great to change the default in an override so it doesn’t make it easy for bloc to configure track while having it default to false.

I basically was trying to have cubit default to true while making it less awkward for classes which extend to cubit to avoid tracking (otherwise you’d see double the transitions).

Normally, I would have totally gone with the suggested approach but it’s a bit of a special case and I’m still brainstorming better/more elegant approaches. Thoughts?

File: packages/cubit/lib/src/cubit.dart (hunk 3 above)
Comment (felangel) [reply packages/cubit/lib/src/cubit.dart]: Yeah good point! I’ll give it a shot haha 😂

File: packages/cubit/lib/src/cubit.dart (hunk 2 above)
Comment (zepfietje) [reply packages/cubit/lib/src/cubit.dart]: Ah, that clarifies the situation!

I don't really see a problem with the following:
```dart
@override
void emit(State state) => super.emit(state, track: false);
```
Or did you mean it feels weird to have to specify `track` in the override?

Comment (felangel) [reply packages/cubit/lib/src/cubit.dart]: I don’t think that’s a valid override of emit haha. The signature should match the superclass.

Comment (zepfietje) [reply packages/cubit/lib/src/cubit.dart]: Whoops, copied it from the bloc codebase which of course doesn't include these `track` changes yet.

Comment (zepfietje) [reply packages/cubit/lib/src/cubit.dart]: Another option would be to make `bool track` nullable (as it currently is) and explicitly check for `track == true` in the `if` statement. Basically the only issue we had now is state changes being tracked when passing `track: null`.

Comment (felangel) [reply packages/cubit/lib/src/cubit.dart]: That was intentional though because I want transitions to be tracked by default. Developers shouldn't need to call `emit(state, track: true)`.

Comment (zepfietje) [reply packages/cubit/lib/src/cubit.dart]: Hmm, yeah I see now.

I can't think of a more elegant approach at the moment.

Comment (felangel) [reply packages/cubit/lib/src/cubit.dart]: The other thing to consider is whether we want to force this notion of events within Cubit in the first place. It might just be best to accept the fact that events don't exist in cubit-land 😄

Comment (tenhobi) [reply packages/cubit/lib/src/cubit.dart]: Good point. From an architecture side of view, Cubit really doesn't need to take care of this. Plus Bloc is only one possible "user" of Cubit. We should let Bloc and other libraries to extend Cubit themselves.

Comment (zepfietje) [reply packages/cubit/lib/src/cubit.dart]: How exactly does this relate to tracking transitions? I suppose we still want to track those, right?

Comment (samandmoore) [reply packages/cubit/lib/src/cubit.dart]: I think that you could still track transitions in Cubit, but without the notion of an event. That would be something that Bloc adds in. 

Also, to accommodate the difference between Cubit and Bloc when it comes to transitions, it might be simpler to split Cubit into two things: a low level custom stream wrapper and a user facing Cubit class. Then Bloc could use the lower level stream wrapper and layer in Events and its own notion of transitions.

File: packages/cubit/lib/src/cubit_stream.dart [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+  /// Returns whether the `Stream<State>` is a broadcast stream.
+  /// Every [CubitStream] is a broadcast stream.
+  @override
+  bool get isBroadcast => _controller.stream.isBroadcast;
+
+  /// Closes the [CubitStream].
+  /// When close is called, new states can no longer be emitted.
+  /// All data on the stream is discarded and a `Future` is returned
+  /// which resolves when it is done or an error ocurred.
```
Comment (zepfietje) [inline packages/cubit/lib/src/cubit_stream.dart]:
```suggestion
  /// which resolves when it is done or an error occurred.
```

File: packages/cubit/test/cubit_test.dart [hunk 6]
```diff
@@ -12,6 +13,43 @@ void main() {
...
+        final cubit = CounterCubit(onTransitionCallback: transitions.add);
+        await Future<void>.delayed(Duration.zero, cubit.increment);
+        await cubit.close();
+        expect(
+          transitions,
+          const [Transition(currentState: 0, nextState: 1)],
+        );
+      });
+
+      test('is called with correct transition for a multiple state changes',
```
Comment (zepfietje) [inline packages/cubit/test/cubit_test.dart]:
```suggestion
      test('is called with correct transitions for multiple state changes',
```

**Review by zepfietje — CHANGES_REQUESTED** [review]
